### ISSUE TITLE
Feature: allow donations thru Paypal

### ISSUE THREAD

**Issue by adrienjoly** [opencollective/opencollective#365]
Justin Jackson makes a great point on his recent article "[Why I bought your software](https://justinjackson.ca/software-saas-product-marketing/)":

> I know PayPal is evil, and a pain in the ass. But for many of us, PayPal is where we keep our play money. It’s a guilt-free spending account.
> 
> Using my credit card feels like a bigger commitment. First, many of us outside of the US don’t have a USD credit card, so we have to pay high conversion rates. Second, the more recurring subscriptions I see on my credit card statement, the more anxiety I have. I don’t get the same stress from PayPal subscriptions.

=> For the same reasons, I believe that it would make a lot of sense to accept Paypal donations on Opencollective.

**Comment by doktor5000** [comment]
>> First, many of us outside of the US don’t have a USD credit card, so we have to pay high conversion rates.

> I believe that it would make a lot of sense to accept Paypal donations on Opencollective.

Same here. I don't have a credit card, but I'm fine donating via PayPal. Certainly not great, but it lowers the entry barrier. And it's probably better to get a donation via PayPal compared to no donation at all, right ?

**Comment by InfiniteMaximum** [comment]
I would love for this to be implemented, if possible. Right now I am trying to find a good funding platform for my gaming clan to support our infrastructure and contest pools. I am looking for a platform that has transparency, pay as you want ,and paypal support. Currently our clan is split between using patreon("You can make a one time donation,but it requires pledging and then cancelling.")or paypal(Not transparent by design). Which currently puts us to using both. This would be the perfect solution if paypal was supported.

**Comment by tonypnode** [comment]
Concur... I would feel much more inclined to donate if my funds came from PayPal rather than from a credit card.

**Comment by piamancini** [comment]
cc / @asood123

**Comment by asood123** [comment]
Good to hear this feedback, it's on my list to bring back.

**Comment by dkebler** [comment]
I don't use my credit cards online if I can avoid it.  Using paypal is actually more secure than typing in ALL your credit card info including cvc so I will wait to see if this happens as being a backer would be nice but is not imperative.  
Not that paypal is really less evil than credit card companies but hey using it may help us get to Mars.   Maybe bitcoins? when you can use finally use them to buy bread and milk at your local store :-).

**Comment by asood123** [comment]
Thanks for keeping this thread active. We plan to work on it in March.

**Comment by MiniDigger** [comment]
since march is coming to an end rather fast now: whats the status on this?

**Comment by clarete** [comment]
I've done some work on researching different PayPal integrations to our system. The latest effort (that won't really get us there yet) was developed here https://github.com/opencollective/opencollective-api/pull/1235.

We might need a different solution on the PayPal side and this depends on their internal approval. But I'll keep this thread updated with the advances in that front. Thank you a lot for your interest in this feature and in open collective!

**Comment by MiniDigger** [comment]
That's strange. I have a recurring payment to Spotify via PayPal without having a credit card added to m account (just two bank accounts). Is this a new restriction?

**Comment by xdamman** [comment]
Mmmm... maybe they just need a backup source of funding?
Or maybe it's ok if you have a balance on your PayPal account?
Is it a US only restriction where credit cards are very widespread (many people in Europe don't have a credit card). Would be good to double check with PayPal.

That said, I still think that there is value to add PayPal even if one would need to attach a credit/debit card. Many people would rather use PayPal instead of typing again their 16 digit credit card number.

**Comment by piamancini** [comment]
@MiniDigger The difference is that in your case one company is charging you directly. In our case we need to charge you on behalf of someone else (the host)

**Comment by astorije** [comment]
@piamancini, would you mind clarifying this please? I am not sure to understand what the host is in this situation. In @MiniDigger's example, it seemed to me that Open Collective charging backer == Spotify charging user.

I agree, @xdamman, having partial PayPal support would be better than none IMO. Another partial support could be to accept one-time donations with PayPal.
I understand that it could push people to go for one-time donations instead of sustainable, recurring donations, but in our case, I could notice that not having PayPal hurts us even more. Besides, so far we have had more one-time donations than recurring donations (though the data sample is very small so take it with a grain of salt :D).

Thanks all for your help on this 🙏

**Comment by xdamman** [comment]
@piamancini that’s the marketplace concept but we won’t be using that given the fact that it has been impossible to have any reliable support from braintree/Paypal. They don’t seem to care. So our intent here is to use Paypal the good old way and charging directly the user with the Paypal credentials of the host (and therefore we won’t be in the middle). So I really don’t see the difference with the Spotify example. If you have a good contact at Paypal who could give us more details on that credit card constrain, that would be very useful.

**Comment by ShinDarth** [comment]
@asood123 any news about this?

**Comment by xdamman** [comment]
@clarete is working on this.

**Comment by clarete** [comment]
Here's how the UI looks like right now. It isn't pretty at all but the functionality is working. Feedback is more than welcome:

![paypalbutton](https://user-images.githubusercontent.com/1459190/40799653-35197872-64dc-11e8-90bf-7507a420958e.gif)

**Comment by alexanderby** [comment]
Here is a good example of a one-click donation https://donate.mozilla.org/en-US/

**Comment by clarete** [comment]
OK, we walked one step forward and two backward. Braintree didn't give us a production account, so I can't really just integrate the branch I have using their API. The next stab at this is to use the [simplest checkout API](https://developer.paypal.com/docs/integration/direct/express-checkout/integration-jsv4/create-payment/) they have.

**Comment by clarete** [comment]
Alright. Here we go with another round of updates.

I've got a test ready to be done on staging. The relevant PRs are these:

  * API: https://github.com/opencollective/opencollective-api/pull/1323
  * Frontend: https://github.com/opencollective/frontend/pull/494

This introduces a new payment provider type called `payment` under the `paypal` payment provider.

Although this is totally functional. I still have one [guard in the code to only allow site admins](https://github.com/opencollective/frontend/pull/494/files#diff-dd61538edecf7711573880aa6a119210R190) (OC people) to use it for now. I did that so we can release it and test it before we open to the broad public.



Is that OK to merge the two PRs above although none of the PRs have any new tests for the feature? I just wanted to make sure that the concept works before I put more effort into it.

**Comment by xdamman** [comment]
Master should always be deployable to production in case we have to apply a hot fix.
Let's instead deploy those branches to a dev/staging environment for such testing.
@znarf can you assist?

**Comment by clarete** [comment]
@xdamman sorry, I feel like I didn't explain exactly the plan. I wouldn't deploy it straight to prod. I'd certainly deploy and test on staging. I asked if it was fine to deploy to prod *after* that without tests for that specific feature because I didn't want to do that before I ensured that it was what we wanted.

**Comment by clarete** [comment]
Here I am once again. I wrote tests (not extensively) for the API side of things. Next step is to write tests to the front-end.

**Comment by clarete** [comment]
It was a lil more complicated to write the tests for the UI since I have a safe guard that only allows admins to use the feature for now. To make sure I'm not breaking anything, I tested it locally and on staging. And this was the scope of the testing:

 * do one purchase with each payment method (stripe and paypal)
 * make sure the new payment method isn't showing up to any collective that isn't hosted by 11004

I did roll it out to production and successfully placed an order. I'm preparing a list of things that we should do before rolling it out to everyone.

![paypaldonation](https://user-images.githubusercontent.com/1459190/41180130-2ec2554c-6b3b-11e8-88d9-8e0ba57c005d.jpg)

**Comment by clarete** [comment]
> Can you also test the recurring part?

Unfortunately not. The API that I used didn't allow me to save any tokens for future payments. I can certainly keep working on it.

**Comment by clarete** [comment]
Just one thing that I just thought about while talking to Pia. I want to check if there is currently any way for Open Collective to easily assess how much money the Open Source Collective retained for a period of time. I'll work on it as soon as this feature lands.

**Comment by clarete** [comment]
Hi @piamancini, thanks for taking the time to test the feature! Here's what I gathered from your feedback:

  1. Allow the user to confirm after clicking the paypal checkout button and before being charged.
  2. Update amount after clicking on PayPal doesn't work
  3. Shouldn't show PayPal unless it's a one time donation

In order to address the problems 1 & 2, I changed a little bit the experience after user clicks "Checkout with PayPal". It will render a confirmation message instead of rendering the order form. Here's how it looks like:

![paypal](https://user-images.githubusercontent.com/1459190/41557332-5352548e-730b-11e8-9063-1e5a2ed4729a.gif)

Ps.: The confirmation page is a placeholder. I'd love to know if you or @xdamman have any suggestions about how it should look like.

**Comment by xdamman** [comment]
If it’s a one time donation, ideally the last step should be to click on the Paypal button and the rest should happen on Paypal website (which should show the amount). After the user confirms on Paypal website, the user should be redirected to their user profile on OC with their latest donation highlighted (exactly as if they would have paid with credit card)

Now for recurring donations, the flow in your comment works but it would be good to be able to show the amount that the user is authorizing for on Paypal. I’m surprised it doesn’t show up there. Or maybe I’m missing something in the GIF?

**Comment by clarete** [comment]
Hi @xdamman thanks for the comment! Let me elaborate a lil more on what's going on:

Before Pia tested it today, this feature was working like you suggested. The user filled in the order details in our form, then clicked the "check out with paypal" button and after filling in their PayPal credentials, they'd be redirected back to the order form that automatically submitted the form.

When @piamancini tried the feature today, she noticed that after signing in with her PayPal account, the form displayed a message stating the following:

 > **"You’ll be able to review your order before you complete your purchase."**

Because of that I started updating the code to show the confirmation page instead of just submitting the order. I think that'd have to happen for both one time donations and subscriptions (for now subscriptions won't work either way so I'm not really doing too much to address problems related to it).

I just slapped that confirmation page together and indeed I think the amount could be a lil easier to see (right now it's a tiny little comment in the disclaimer underneath the confirmation button). I'm working on a better version of it. Just wanted to make sure I was in the right direction before got too much work done!

**Comment by piamancini** [comment]
@xdamman the paypal modal wasn't showing up the amount and since the UI for payment is not super clear I think the extra confirmation is needed. 
As an example: I thought I was making a $10 donation but made a $50 one. At no point in the process I saw the $50 (except on the order form that's hard to miss)

**Comment by clarete** [comment]
I've got a slightly better UI for the review page. Let me know if that works or if I'm missing something:

![paypal-order-review](https://user-images.githubusercontent.com/1459190/41562118-b47dc55a-7319-11e8-8d93-e956c914d162.jpg)

**Comment by xdamman** [comment]
Looks like the problem is in making sure that the Paypal website shows the amount. Which it’s not the case in the gif above for whatever reason. 

The process of connecting your Paypal account before making the payment is really weird. I’ve never seen it before. Can you point me to a website that does it that way?

For one time payments, I always expect to make the actual payment on Paypal.com.

What is actually happening here in this process behind the scene? The Paypal website isn’t super clear. Is the user authorizing a max amount per charge? Recurring? That will help me figure out the best process.

**Comment by xdamman** [comment]
Your order review screen looks great 👍 
We should probably breakdown the order flow in multiple steps. Will make things easier.

I’m fine with keeping the Paypal flow that way. It’s a bit weird for one time payments because it’s not what people are used to but it’s not the end of the world.

**Comment by xdamman** [comment]
Maybe just add the host receiving the money on behalf of the collective to that screen

**Comment by clarete** [comment]
Here's an update on what we've done offline. @xdamman pointed me to the flag `commit` in the PayPal button API that allows users to checkout from within PayPal instead of having to see a confirmation page. With that option set, the order review screen ended up being useless so we took it out.

After we tested the above change, I removed the restriction that made PayPal available only for platform admins. Now everyone can use PayPal for one time donations to collectives hosted by the Open Source Collective!!

I wonder if I should close this ticket and maybe open a new one to track the work on the recurring payment.
